### PR TITLE
Update `connectivity` filter with new extraction modes and scalars

### PR DESCRIPTION
**Pull request by user27182**
### Overview

This PR updates the [`connectivity` filter](https://docs.pyvista.org/version/0.41/api/core/_autosummary/pyvista.DataSetFilters.connectivity.html) to include additional functionality provided by the underlying [`vtkConnectivityFilter`](https://vtk.org/doc/nightly/html/classvtkConnectivityFilter.html).

### Details

- Add parameter `extraction_mode` to control overall filter behaviour. Existing modes are `'all'` and `'largest'`, new modes are `'specified'`, `'cell_seed'`, `'point_seed'`, `'closest'`
- Deprecate parameter `largest=True` and replace it with `extraction_mode='largest'`
- Add sorting output regions by default (previously unsorted/random ordering)
- Add parameters `scalar_range` and `scalars` to control / restrict connectivity
- Add parameters `region_ids`, `cell_ids`, `point_ids`, `closest_point` to make use of `'specified'`, `'cell_seed'`, `'point_seed'`, `'closest'` extraction modes, respectively
- Add parameter `variable_input` to simplify the API and remove the need to explicitly use mode-specific parameters by name
- Add `label_regions` parameter to control labelling with `RegionId` scalars
- Add `inplace` parameter
- Add new examples showcasing all extraction modes
- Update `connectivity` and `compute-volume` reference examples
- Improve filter documentation
- Generalize filter to better (i.e. correctly) handle different dataset input types
- Add extensive filter testing

### Compatibility
- Apart from deprecating the parameter `largest=True`, the existing usage of the filter is unchanged
- Some minor behaviour changes:
(1) The `RegionId` scalars of `connectivity('all')` are now sorted by default, whereas they previously were not sorted at all, so the new scalar values are different from before
(2) Before, `extract_largest()` was equivalent to `connectivity(largest=True)`, but since `vtkConnectivityFilter` is buggy (see #4764 for an example), it was necessary for the filter to remove the `RegionId` scalars from the output. With this update, however, the issue with the scalars is fixed, and so the default output of `connectivity('largest')` has scalars where it previously did not. And so, `extract_largest()` is now no longer equivalent to `connectivity('largest')`, since the former has `label_regions=False` by default whereas the latter has `label_regions=True` by default. This is relevant because the `connectivity` and `compute-volume` reference examples both previously mentioned that the two methods were equivalent, but now they technically are not.

### Implementation details
- The underlying `vtkConnectivityFilter` is _very_ buggy and was found to have an inconsistent/unfaithful API that would regularly produce unexpected output
- There were issues with pretty much every input dataset type, extraction mode, region labels, and scalar values. As a result, a number of deviations / corrections needed to be made to make the API consistent and ensure correct output... in some cases this has a performance penalty 
- e.g. if the output scalars do not match the size of the points/cells, the filter is called recursively a second time to fix the output
- e.g. the `'specified'` mode would not sort the output correctly, so instead all regions are extracted and then `extract_points` is used afterward to manually filter the output
- e.g. scalar range does not work for `'all'`, and so it was necessary to use `extract_points` to manually restrict the cell connectivity before applying the filter
- In cases where `extract_points` is used on `PolyData` input, the output `UnstructuredGrid` was converted back to `PolyData` using `vtkRemovePolyData`, this was done to stay faithful to the vtkConnectivityFilter which promises `PolyData` output for `PolyData` input

### CODE REVIEW

**Comment by akaszynski** [comment]
Really great work here and excellent docs:

tmp

Made several edits to the docs to get the style to build correctly. Note how our docstring examples can't use rst sections, but we can use `**` instead. The example section might be a candidate for a fully fledged pyvista example, but I'm fine if you want to hold off from that as the existing example looks fine in the rendered docs.

Also, made a few other changes including `TypeError` --> `ValueError` for unspecified kwargs for certain `extraction_mode`.

File: pyvista/core/_vtk_core.py [hunk 1]
```diff
@@ -265,6 +270,7 @@
     vtkOBBTree,
     vtkRectilinearGridToPointSet,
     vtkRectilinearGridToTetrahedra,
+    vtkRemovePolyData,
```
Comment (tkoyama010) [inline pyvista/core/_vtk_core.py:273]:
```suggestion
```

File: pyvista/core/filters/data_set.py [hunk 2]
```diff
@@ -2373,58 +2373,479 @@ def glyph(
...
-        criterion such as scalar range are known as a region.
+        In general, cells are considered to be connected if they
+        share a point. However, if a ``scalar_range`` is provided, cells
+        must also have at least one point with scalar values in the
+        specified range to be considered connected.
+
+        See :ref:`connectivity_example` and :ref:`volumetric_example` for
+        more examples using this filter.
+
+        .. versionadded:: 0.42.0
```
Comment (tkoyama010) [inline pyvista/core/filters/data_set.py:2408]:
```suggestion
        .. versionadded:: 0.43.0
```

File: pyvista/core/filters/data_set.py [hunk 3]
```diff
@@ -2373,58 +2373,479 @@ def glyph(
...
+           * New extraction modes: ``'specified'``, ``'cell_seed'``, ``'point_seed'``,
+             and ``'closest'``.
+           * Extracted regions are now sorted in descending order by
+             cell count.
+           * Region connectivity can be controlled using ``scalar_range``.
+
+        .. deprecated:: 0.42.0
```
Comment (tkoyama010) [inline pyvista/core/filters/data_set.py:2415]:
```suggestion
        .. deprecated:: 0.43.0
```

File: pyvista/core/filters/data_set.py [hunk 4]
```diff
@@ -2373,58 +2373,479 @@ def glyph(
...
+        >>> # Use a categorical colormap
+        >>> categories = True
+        >>> cmap = 'glasbey'
+        >>>
+        >>> # Format scalar bar text for integer values
+        >>> scalar_bar_args = dict(
+        ...     fmt='%.f',  # Do not show decimals
+        ... )
+        >>>
+        >>> # Set the plot's camera position
```
Comment (tkoyama010) [inline pyvista/core/filters/data_set.py:2529]:
```suggestion

        Set the plot's camera position
```

File: pyvista/core/filters/data_set.py [hunk 5]
```diff
@@ -2373,58 +2373,479 @@ def glyph(
...
+        >>> cpos = [(10.5, 12.2, 18.3), (0.0, 0.0, 0.0), (0.0, 1.0, 0.0)]
+
+        Plot the regions by ID.
+
+        >>> # Note that region IDs are sorted in descending order by
+        >>> # cell count
```
Comment (tkoyama010) [inline pyvista/core/filters/data_set.py:2535]:
```suggestion

        Note that region IDs are sorted in descending order by
        cell count
```

File: pyvista/core/filters/data_set.py [hunk 6]
```diff
@@ -2373,58 +2373,479 @@ def glyph(
...
+        ...     cmap=cmap,
+        ...     scalar_bar_args=scalar_bar_args,
+        ...     cpos=cpos,
+        ... )
+
+        **Extract specific regions by size**
+
+        Calculate region sizes.
+
+        >>> # Get counts using the previous `connectivity('all')` results
```
Comment (tkoyama010) [inline pyvista/core/filters/data_set.py:2547]:
```suggestion
        Get counts using the previous `connectivity('all')` results
```

File: pyvista/core/filters/data_set.py [hunk 7]
```diff
@@ -2373,58 +2373,479 @@ def glyph(
...
+        ...     cmap=cmap,
+        ...     scalar_bar_args=scalar_bar_args,
+        ...     cpos=cpos,
+        ... )
+
+        **Extract the largest region**
+
+        >>> conn = mesh.connectivity('largest', label_regions=False)
+
+        >>> # Plot largest region and show input mesh for reference
```
Comment (tkoyama010) [inline pyvista/core/filters/data_set.py:2570]:
```suggestion
        Plot the largest region and show the input mesh for reference
```

File: pyvista/core/filters/data_set.py [hunk 8]
```diff
@@ -2373,58 +2373,479 @@ def glyph(
...
+        >>> p = pv.Plotter()
+        >>> _ = p.add_mesh(conn)
+        >>> _ = p.add_mesh(mesh, style='wireframe')
+        >>> p.show(cpos=cpos)
+
+        **Extract regions using seed points**
+
+        Create hills and use curvature to define its peaks and valleys
```
Comment (tkoyama010) [inline pyvista/core/filters/data_set.py:2578]:
```suggestion
        Create hills and use curvature to define their peaks and valleys
```

File: pyvista/core/filters/data_set.py [hunk 9]
```diff
@@ -2373,58 +2373,479 @@ def glyph(
...
+
+        >>> import pyvista as pv
+        >>> mesh = pv.ParametricRandomHills()
+        >>> mesh["Curvature"] = mesh.curvature()
+
+        Visualize the peaks and valleys
+
+        >>> # Peaks have large positive curvature (i.e. are convex)
+        >>> # Valleys have large negative curvature (i.e. are concave)
+        >>> # Flat regions have curvature close to zero
```
Comment (tkoyama010) [inline pyvista/core/filters/data_set.py:2588]:
```suggestion
        Peaks have large positive curvature (i.e. are convex)
        Valleys have large negative curvature (i.e. are concave)
        Flat regions have curvature close to zero

```

File: pyvista/core/filters/data_set.py [hunk 10]
```diff
@@ -2373,58 +2373,479 @@ def glyph(
...
+        >>> _ = p.add_mesh(mesh, style='wireframe', color='lightgray')
+        >>> _ = p.add_mesh(peak_mesh, color='red', label='Steepest Peak')
+        >>> _ = p.add_mesh(
+        ...     valley_mesh, color='blue', label='Closest Valley'
+        ... )
+        >>> _ = p.add_legend()
+        >>> p.show()
 
         """
+        # Deprecated on v0.42.0
```
Comment (tkoyama010) [inline pyvista/core/filters/data_set.py:2625]:
```suggestion
        # Deprecated on v0.43.0
```

File: pyvista/core/filters/data_set.py [hunk 11]
```diff
@@ -2373,58 +2373,479 @@ def glyph(
...
+            is_all_integers = np.issubdtype(ids.dtype, np.integer)
+            is_all_positive = not np.any(ids < 0)
+            if not (is_all_positive and is_all_integers):
+                raise ValueError('IDs must be positive integer values.')
+            return np.unique(ids)
+
+        def _extract_points(_input, points):
+            # This method is similar to `mesh.extract_points` but the
+            # output arrays are removed and the output type is PolyData
+            # if input type is PolyData
```
Comment (tkoyama010) [inline pyvista/core/filters/data_set.py:2646]:
```suggestion
            # This method is similar to `mesh.extract_points` but the
            # output arrays are removed and the output type is PolyData
            # if the input type is PolyData
```

File: pyvista/core/filters/data_set.py [hunk 12]
```diff
@@ -2373,58 +2373,479 @@ def glyph(
...
+            _output = _input.extract_points(points, progress_bar=progress_bar)
+            if isinstance(_input, pyvista.PolyData):
+                # Output is UnstructuredGrid, so apply vtkRemovePolyData
+                # to input to make output as PolyData type instead
```
Comment (tkoyama010) [inline pyvista/core/filters/data_set.py:2650]:
```suggestion
                # Output is UnstructuredGrid, so apply vtkRemovePolyData
                # to input to make the output as PolyData type instead
```

File: pyvista/core/filters/data_set.py [hunk 13]
```diff
@@ -2373,58 +2373,479 @@ def glyph(
...
+                    upper = input_mesh.active_scalars <= scalar_range[1]
+                    input_mesh = _extract_points(input_mesh, np.logical_and(upper, lower))
+
         alg = _vtk.vtkConnectivityFilter()
-        alg.SetInputData(self)
-        if largest:
+        alg.SetInputDataObject(input_mesh)
+
+        # Due to inconsistent / buggy output, always keep this on and
+        # remove scalars later as needed
```
Comment (tkoyama010) [inline pyvista/core/filters/data_set.py:2717]:
```suggestion
        # Due to inconsistent/buggy output, always keep this on and
        # remove scalars later as needed
```

File: tests/core/test_dataset_filters.py [hunk 14]
```diff
@@ -1003,7 +1003,343 @@ def test_glyph_orient_and_scale():
     assert glyph4.bounds[0] == geom.bounds[0] and glyph4.bounds[1] == geom.bounds[1]
 
 
-def test_split_and_connectivity():
+@pytest.fixture
+def connected_datasets():
+    # This is similar to datasets fixture, but the PolyData is fully connected
```
Comment (tkoyama010) [inline tests/core/test_dataset_filters.py:1008]:
```suggestion
    # This is similar to the datasets fixture, but the PolyData is fully connected
```

File: tests/core/test_dataset_filters.py [hunk 15]
```diff
@@ -1003,7 +1003,343 @@ def test_glyph_orient_and_scale():
...
+    'extraction_mode', ['all', 'largest', 'specified', 'cell_seed', 'point_seed', 'closest']
+)
+@pytest.mark.parametrize('label_regions', [True, False])
+@pytest.mark.parametrize('scalar_range', [True, False])
+@pytest.mark.needs_vtk_version(9, 1, 0)
+def test_connectivity_inplace_and_output_type(
+    datasets, dataset_index, extraction_mode, label_regions, scalar_range
+):
+    # paramaterize with label_regions and scalar_range as these parameters
+    # have branches which may modify input / input type
```
Comment (tkoyama010) [inline tests/core/test_dataset_filters.py:1073]:
```suggestion
    # paramaterize with label_regions and scalar_range as these parameters
    # have branches which may modify input/input type
```

**Review by tkoyama010 — APPROVED** [review]
LGTM. Sorry for being late.

File: pyvista/core/filters/data_set.py [hunk 16]
```diff
@@ -2373,58 +2373,485 @@ def glyph(
...
+        Plot result with plotting parameters defined earlier.
+
+        >>> conn.plot(
+        ...     categories=categories,
+        ...     cmap=cmap,
+        ...     scalar_bar_args=scalar_bar_args,
+        ...     cpos=cpos,
+        ... )
+
+        **Extract the largest region**
```
Comment (tkoyama010) [inline pyvista/core/filters/data_set.py:2570]:
```suggestion
        Extract the largest region.
```

File: pyvista/core/filters/data_set.py [hunk 17]
```diff
@@ -2373,58 +2373,485 @@ def glyph(
...
+
+        >>> conn = mesh.connectivity('largest', label_regions=False)
+
+        Plot the largest region and show the input mesh for reference
```
Comment (tkoyama010) [inline pyvista/core/filters/data_set.py:2574]:
```suggestion
        Plot the largest region and show the input mesh for reference.
```

File: pyvista/core/filters/data_set.py [hunk 18]
```diff
@@ -2373,58 +2373,485 @@ def glyph(
...
+
+        >>> p = pv.Plotter()
+        >>> _ = p.add_mesh(conn)
+        >>> _ = p.add_mesh(mesh, style='wireframe')
+        >>> p.show(cpos=cpos)
+
+        **Extract regions using seed points**
```
Comment (tkoyama010) [inline pyvista/core/filters/data_set.py:2581]:
```suggestion
        Extract regions using seed points.
```

File: pyvista/core/filters/data_set.py [hunk 19]
```diff
@@ -2373,58 +2373,485 @@ def glyph(
...
+
+        >>> mesh.plot(
+        ...     clim=[-0.5, 0.5],
+        ...     cmap='bwr',
+        ...     below_color='blue',
+        ...     above_color='red',
+        ... )
 
-        See :ref:`volumetric_example` for more examples using this filter.
+        Extract the steepest peak using a seed point
```
Comment (tkoyama010) [inline pyvista/core/filters/data_set.py:2602]:
```suggestion
        Extract the steepest peak using a seed point.
```

File: pyvista/core/filters/data_set.py [hunk 20]
```diff
@@ -2373,58 +2373,485 @@ def glyph(
...
+
+        >>> data_min, data_max = mesh.get_data_range()
+        >>> peak_range = [0.2, data_max]  # Peak if curvature > 0.2
+        >>> peak_point_id = np.argmax(mesh['Curvature'])  # Steepest point
+        >>> peak_mesh = mesh.connectivity(
+        ...     'point_seed', peak_point_id, scalar_range=peak_range
+        ... )
+
+        Extract the valley region closest to the steepest peak
```
Comment (tkoyama010) [inline pyvista/core/filters/data_set.py:2611]:
```suggestion
        Extract the valley region closest to the steepest peak.
```

**Review by tkoyama010 — DISMISSED** [review]

File: pyvista/core/filters/data_set.py [hunk 21]
```diff
@@ -2373,58 +2373,485 @@ def glyph(
...
+            * ``'point_seed'`` : Extract all regions sharing the specified
+              point ids. Use ``point_ids``to specify the point ids.
+            * ``'closest'`` : Extract the region closest to the specified
+              point. Use ``closest_point`` to specify the point.
+
+        variable_input : float | sequence[float], optional
+            The convenience parameter used for specifying any required input values
+            for some values of ``extraction_mode``. Setting
+            ``extraction_input`` is equivalent to setting:
+            * ``'region_ids'`` if mode is ``'specified'``
```
Comment (tkoyama010) [inline pyvista/core/filters/data_set.py:2438]:
```suggestion
            * ``'region_ids'`` if mode is ``'specified'``.
```

File: pyvista/core/filters/data_set.py [hunk 22]
```diff
@@ -2373,58 +2373,485 @@ def glyph(
...
+            * ``'cell_ids'`` if mode is ``'cell_seed'``
```
Comment (tkoyama010) [inline pyvista/core/filters/data_set.py:2439]:
```suggestion
            * ``'cell_ids'`` if mode is ``'cell_seed'``.
```

File: pyvista/core/filters/data_set.py [hunk 23]
```diff
@@ -2373,58 +2373,485 @@ def glyph(
...
+            * ``'point_ids'`` if mode is ``'point_seed'``
```
Comment (tkoyama010) [inline pyvista/core/filters/data_set.py:2440]:
```suggestion
            * ``'point_ids'`` if mode is ``'point_seed'``.
```

File: pyvista/core/filters/data_set.py [hunk 24]
```diff
@@ -2373,58 +2373,485 @@ def glyph(
...
+            ``pyvista.UnstructuredGrid`` otherwise.
+
+        See Also
+        --------
+            :func:`pyvista.DataSetFilter.connectivity`
 
         Examples
         --------
-        Join two meshes together and plot their connectivity.
+        **Label all connected regions**
```
Comment (tkoyama010) [inline pyvista/core/filters/data_set.py:2505]:
```suggestion
        Label all connected regions.
```

File: pyvista/core/filters/data_set.py [hunk 25]
```diff
@@ -2373,58 +2373,485 @@ def glyph(
...
+        >>> import pyvista as pv
+        >>> mesh = examples.download_foot_bones()
+
+        Extract all disconnected regions.
+
+        >>> conn = mesh.connectivity()
+
+        Configure plotting parameters.
+
+        >>> # Use a categorical colormap
```
Comment (tkoyama010) [inline pyvista/core/filters/data_set.py:2520]:
```suggestion
```

File: pyvista/core/filters/data_set.py [hunk 26]
```diff
@@ -2373,58 +2373,485 @@ def glyph(
...
+        >>> categories = True
+        >>> cmap = 'glasbey'
+
+        Format scalar bar text for integer values
```
Comment (tkoyama010) [inline pyvista/core/filters/data_set.py:2524]:
```suggestion
        Format scalar bar text for integer values.
```

File: pyvista/core/filters/data_set.py [hunk 27]
```diff
@@ -2373,58 +2373,485 @@ def glyph(
...
+
+        >>> scalar_bar_args = dict(
+        ...     fmt='%.f',  # Do not show decimals
+        ... )
+
+        Set the plot's camera position
```
Comment (tkoyama010) [inline pyvista/core/filters/data_set.py:2530]:
```suggestion
        Set the plot's camera position.
```

File: pyvista/core/filters/data_set.py [hunk 28]
```diff
@@ -2373,58 +2373,485 @@ def glyph(
...
+
+        >>> cpos = [(10.5, 12.2, 18.3), (0.0, 0.0, 0.0), (0.0, 1.0, 0.0)]
+
+        Plot the regions by ID.
+
+        Note that region IDs are sorted in descending order by
+        cell count
```
Comment (tkoyama010) [inline pyvista/core/filters/data_set.py:2537]:
```suggestion
        cell count.
```

File: pyvista/core/filters/data_set.py [hunk 29]
```diff
@@ -2373,58 +2373,485 @@ def glyph(
...
+
+        >>> conn.plot(
+        ...     categories=categories,
+        ...     cmap=cmap,
+        ...     scalar_bar_args=scalar_bar_args,
+        ...     cpos=cpos,
+        ... )
+
+        **Extract specific regions by size**
```
Comment (tkoyama010) [inline pyvista/core/filters/data_set.py:2546]:
```suggestion
        Extract specific regions by size.
```

File: pyvista/core/filters/data_set.py [hunk 30]
```diff
@@ -2373,58 +2373,485 @@ def glyph(
...
+
```
Comment (tkoyama010) [inline pyvista/core/filters/data_set.py:2547]:
```suggestion
```

File: pyvista/core/filters/data_set.py [hunk 31]
```diff
@@ -2373,58 +2373,485 @@ def glyph(
...
+        Calculate region sizes.
+
+        Get counts using the previous `connectivity('all')` results
```
Comment (tkoyama010) [inline pyvista/core/filters/data_set.py:2550]:
```suggestion
        Get counts using the previous `connectivity('all')` results.
```

File: pyvista/core/filters/data_set.py [hunk 32]
```diff
@@ -2373,58 +2373,485 @@ def glyph(
...
+        Plot the largest region and show the input mesh for reference.
+
+        >>> p = pv.Plotter()
+        >>> _ = p.add_mesh(conn)
+        >>> _ = p.add_mesh(mesh, style='wireframe')
+        >>> p.show(cpos=cpos)
+
+        Extract regions using seed points.
+
+        Create hills and use curvature to define their peaks and valleys
```
Comment (tkoyama010) [inline pyvista/core/filters/data_set.py:2583]:
```suggestion
        Create hills and use curvature to define their peaks and valleys.
```

File: pyvista/core/filters/data_set.py [hunk 33]
```diff
@@ -2373,58 +2373,485 @@ def glyph(
...
+
+        >>> import pyvista as pv
+        >>> mesh = pv.ParametricRandomHills()
+        >>> mesh["Curvature"] = mesh.curvature()
+
+        Visualize the peaks and valleys
```
Comment (tkoyama010) [inline pyvista/core/filters/data_set.py:2589]:
```suggestion
        Visualize the peaks and valleys.
```

File: pyvista/core/filters/data_set.py [hunk 34]
```diff
@@ -2373,58 +2373,485 @@ def glyph(
...
+
+        Peaks have large positive curvature (i.e. are convex)
+        Valleys have large negative curvature (i.e. are concave)
+        Flat regions have curvature close to zero
```
Comment (tkoyama010) [inline pyvista/core/filters/data_set.py:2593]:
```suggestion
        Flat regions have curvature close to zero.
```

File: pyvista/core/filters/data_set.py [hunk 35]
```diff
@@ -2373,58 +2373,483 @@ def glyph(
...
-        largest : bool, default: False
-            Extract the largest connected part of the mesh.
+        extraction_mode : str, default: "all"
+            * ``'all'``: Extract all connected regions.
+            * ``'largest'`` : Extract the largest connected region (by cell
+              count).
+            * ``'specified'``: Extract specific region IDs. Use ``region_ids``
+              to specify the region IDs to extract.
+            * ``'cell_seed'``: Extract all regions sharing the specified cell
+              ids. Use ``cell_ids``to specify the cell ids.
```
Comment (tkoyama010) [inline pyvista/core/filters/data_set.py:2428]:
```suggestion
              ids. Use ``cell_ids`` to specify the cell ids.
```

File: pyvista/core/filters/data_set.py [hunk 36]
```diff
@@ -2373,58 +2373,483 @@ def glyph(
...
+            * ``'point_seed'`` : Extract all regions sharing the specified
+              point ids. Use ``point_ids``to specify the point ids.
```
Comment (tkoyama010) [inline pyvista/core/filters/data_set.py:2430]:
```suggestion
              point ids. Use ``point_ids`` to specify the point ids.
```

**Review by tkoyama010 — APPROVED** [review]

**Pull request merged**